docs/fsharp/style-guide/component-design-guidelines.md

```diff
@@ -693,11 +693,83 @@ let checkNonNull argName (arg: obj) =
     | null -> nullArg argName
     | _ -> ()
 
-let checkNonNull` argName (arg: obj) =
+let checkNonNull' argName (arg: obj) =
     if isNull arg then nullArg argName
     else ()
 ```
 
+Starting with F# 9, you can leverage the new `| null` [syntax](../language-reference/values/null-values.md#null-values-starting-with-f-9) to make the compiler indicate possible null values and where they need handling. The above code will then start producing warnings:
+
+```fsharp
+let checkNonNull argName (arg: obj) =
+    match arg with
+    // nullness warning: the type 'obj' does not support 'null'
+    | null -> nullArg argName
+    | _ -> ()
+
+let checkNonNull' argName (arg: obj) =
+    // nullness warning: the type 'obj' does not support 'null'
+    if isNull arg then nullArg argName 
+    else ()
+```
+
+To address the warnings, you could explicitly specify `null` as a possible argument value:
+
+```fsharp
+let checkNonNull argName (arg: obj | null) =
+    match arg with
+    | null -> nullArg argName
+    | _ -> ()
+
+let checkNonNull' argName (arg: obj | null) =
+    if isNull arg then nullArg argName 
+    else ()
+```
+
+Yet, the most succinct way to achieve this is to use the new builtin `nullArgCheck` function:
+
+```fsharp
+let checkNonNull'' argName arg =    // `arg` is inferred to be `obj | null`
+    nullArgCheck argName arg
+    |> ignore
+```
+
+The `nullArgCheck` function is also handy when you need to sanitize the input and keep it in case it's not null:
+
+```fsharp
+let processNullableList l =       
+    // `l` is inferred as `int list | null`
+    let l = nullArgCheck (nameof l) l
+    // now, `l` is inferred as `int list` and is safe to use
+    l |> List.map (fun x -> x + 1)
+```
+
+On the other hand, you'll get a warning if the compiler detects that a possible null value is not handled:
+
+```fsharp
+let printLineLength (s: string) =
+    printfn "%i" s.Length
+
+let readLineFromStream (sr: System.IO.StreamReader) =
```

Comment:
This is a good example and might as well come first.
It should say where the nulls comes from - it is part of the annotation for `.ReadLine()`

Reply:
Right, added a note.